### PR TITLE
Run static analysis  using  PHPStan on CI

### PR DESCRIPTION
**Pull request by swissspidy**
This already helped catch a few bugs.

See #319.

### CODE REVIEW

**Comment by swissspidy** [comment]
@szepeviktor Hope you don't mind this ping! It'd be awesome if you could take a look at this PR and let me know your thoughts about how  I've implemented `szepeviktor/phpstan-wordpress` here.

File: .travis.yml [hunk 1]
```diff
@@ -28,7 +28,7 @@ cache:
     - $HOME/.jest-cache
     - $HOME/.npm
     - $HOME/.nvm/.cache
-    - $HOME/.wp-cli/.cache
+    - $HOME/.wp-cli/.cachec
```
Comment (szepeviktor) [inline .travis.yml]: @swissspidy Is there a stray `c` here?

**Comment by szepeviktor** [comment]
This is definitely a nice PR.

# :-)

**Review by szepeviktor — APPROVED** [review]

**Comment by spacedmonkey** [comment]
This didnt install locally 
```
~/workspace/web-stories-wp  add/phpstan ✔                                                                                                                                                                                                     1h46m  ⍉
▶ composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for phpstan/phpstan 0.12.11 -> satisfiable by phpstan/phpstan[0.12.11].
    - phpstan/phpstan 0.12.11 requires php ^7.1 -> your PHP version (7.4.2) overridden by "config.platform.php" version (5.6) does not satisfy that requirement.
  Problem 2
    - Installation request for szepeviktor/phpstan-wordpress v0.6.0 -> satisfiable by szepeviktor/phpstan-wordpress[v0.6.0].
    - szepeviktor/phpstan-wordpress v0.6.0 requires php ~7.1 -> your PHP version (7.4.2) overridden by "config.platform.php" version (5.6) does not satisfy that requirement.

```

File: .travis.yml [hunk 2]
```diff
@@ -46,7 +46,13 @@ install:
     fi
   - |
     if [ "$LINT" = "1" ] || [ "$PHP" = "1" ] || [ "$DEPLOY_PLUGIN" = "1" ]; then
-      composer install
+      composer install -n --ignore-platform-reqs
```
Comment (spacedmonkey) [inline .travis.yml]: Why not add `--no-interaction --no-suggest` here?

File: .travis.yml [hunk 3]
```diff
@@ -46,7 +46,13 @@ install:
...
+    fi
+  - |
+    if [ $(phpenv version-name) = "5.6" ] || [ $(phpenv version-name) = "7.0" ]; then
+      # phpstan only supports PHP 7.1+. Ignore version conflicts on lower versions.
+      composer remove -n --dev szepeviktor/phpstan-wordpress --ignore-platform-reqs
+      composer update -n --optimize-autoloader
```
Comment (spacedmonkey) [inline .travis.yml]: Why not add `--no-interaction --no-suggest` here?

File: .travis.yml [hunk 4]
```diff
@@ -46,7 +46,13 @@ install:
     fi
   - |
     if [ "$LINT" = "1" ] || [ "$PHP" = "1" ] || [ "$DEPLOY_PLUGIN" = "1" ]; then
-      composer install
+      composer install -n --ignore-platform-reqs
+    fi
+  - |
+    if [ $(phpenv version-name) = "5.6" ] || [ $(phpenv version-name) = "7.0" ]; then
```
Comment (spacedmonkey) [inline .travis.yml]: So do run in on PHP 5.6 or 7?

**Comment by spacedmonkey** [comment]
This looks cool. But I do not think it should run with all other lints. The analysis from this should not run on every PHP type. It doesn't make a lot of sense. This should be another job that is run in parallel, only on PHP 7.4 ( for speed and compatibility ). Maybe named analysis. In future we can add more to analysis, checking javascript and other things.

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Changes requested.

File: .travis.yml (hunk 3 above)
Comment (szepeviktor) [reply .travis.yml]: `-n` is `--no-interaction` - although not nice to be short

File: .travis.yml (hunk 2 above)
Comment (swissspidy) [reply .travis.yml]: `-n` is short for `--no-interaction`. I do not see how `--no-suggest` is relevant here.

**Comment by swissspidy** [comment]
> This didnt install locally

Yeah it needs `--ignore-platform-reqs` because it's PHP 7.1+ only. I might just add it to Travis CI only and remove it from `composer.json` because of this.

> The analysis from this should not run on every PHP type. It doesn't make a lot of sense. This should be another job that is run in parallel, only on PHP 7.4 ( for speed and compatibility )

What do you mean with  every  PHP type? There is just one lint job that runs on PHP 7.4, which this is part of.

File: .travis.yml (hunk 1 above)
Comment (swissspidy) [reply .travis.yml]: Whoops, indeed! No idea how that got there. Thanks!

File: .travis.yml (hunk 4 above)
Comment (swissspidy) [reply .travis.yml]: This part here means: do not install / do remove phpstan on PHP <  7.1.

It only runs on 7.4 with the lint job.

But as per my other comment I might change this so it becomes less complicated.

File: .travis.yml (hunk 3 above)
Comment (swissspidy) [reply .travis.yml]: Why not nice? Not clear what it does?

Comment (szepeviktor) [reply .travis.yml]: If someone sees `--no-interaction` and speaks a little bit of English he will know what it is for, `-n` is not like that!

**Comment by szepeviktor** [comment]
```yaml
    - name: Static Analysis (PHPStan)
      php: 7.4
      env: ANALYSIS=1
```

File: .travis.yml (hunk 2 above)
Comment (szepeviktor) [reply .travis.yml]: `--no-suggest` hides all the suggestions packages make - it makes CI output more readable

Comment (swissspidy) [reply .travis.yml]: I know, but suggestions only make up for  about 4 lines in CI output that is collapsed anyway. Don't see the immediate need to change anything here.

**Comment by swissspidy** [comment]
@szepeviktor What's your comment about?

**Comment by szepeviktor** [comment]
To show that it runs only once.

**Comment by swissspidy** [comment]
Gotcha. I changed that after the feedback you two shared. But it ran only once before too :-)

**Review by spacedmonkey — APPROVED** [review]
👍 Looks really good.

**Comment by szepeviktor** [comment]
It looks good on my resume:

> Google uses my software.

**Comment by ondrejmirtes** [comment]
Awesome to see that Google started using PHPStan :) Let me know if you need any help.

I already have a suggestion - you could run a higher level (like level 6) and introduce a baseline so that you don't have to fix all the found issues in existing code at once, but have them be pointed out for new code. Read more about it here: https://medium.com/@ondrejmirtes/phpstans-baseline-feature-lets-you-hold-new-code-to-a-higher-standard-e77d815a5dff

**Comment by swissspidy** [comment]
Thanks for chiming in, @ondrejmirtes! 🙂 
I figured I'd give it a go for this project. Didn't pay too much attention to the various levels, but bumping it was easy, see #350.

**Pull request merged**